src/tools/fuzzing/fuzzing.cpp

```diff
@@ -1916,76 +1931,143 @@ Literal TranslateToFuzzReader::makeLiteral(Type type) {
   WASM_UNREACHABLE("invalid value");
 }
 
-Expression* TranslateToFuzzReader::makeConst(Type type) {
-  if (type.isRef()) {
-    assert(wasm.features.hasReferenceTypes());
-    // Check if we can use ref.func.
-    // 'funcContext->func' is the pointer to the last created function and can
-    // be null when we set up globals (before we create any functions), in
-    // which case we can't use ref.func.
-    if (type == Type::funcref && funcContext && oneIn(2)) {
-      // First set to target to the last created function, and try to select
-      // among other existing function if possible
-      Function* target = funcContext->func;
-      if (!wasm.functions.empty() && !oneIn(wasm.functions.size())) {
-        target = pick(wasm.functions).get();
-      }
-      return builder.makeRefFunc(target->name, target->type);
-    }
-    if (type == Type::i31ref) {
-      return builder.makeI31New(makeConst(Type::i32));
-    }
-    if (oneIn(2) && type.isNullable()) {
-      return builder.makeRefNull(type);
-    }
-    if (!type.isFunction()) {
-      // We don't know how to create an externref or GC data yet TODO
-      // For now, create a null, and if it must be non-null, cast it to such
-      // even though that traps at runtime.
-      auto nullable = Type(type.getHeapType(), Nullable);
-      Expression* ret = builder.makeRefNull(nullable);
-      if (!type.isNullable()) {
-        ret = builder.makeRefAs(RefAsNonNull, ret);
-      }
-      return ret;
+Expression* TranslateToFuzzReader::makeRefFuncConst(Type type) {
+  // Neither ref.func nor ref.as_non_null are allowed in globals, so we can only
```

Comment:
`ref.func` isn't allowed in globals? j2wasm vtables and itables use them. Is it just at the top level that `ref.func` isn't allowed? That seems odd if so.

Reply:
Actually, I think this is just a temporary restriction on the fuzzer, since we create globals before we create functions. I'll clarify the comment and look at lifting this restriction in a follow-up.